tls/s2n_x509_validator.c

```diff
@@ -213,95 +214,159 @@ int s2n_x509_validator_set_max_chain_depth(struct s2n_x509_validator *validator,
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(current_name);
+    RESULT_ENSURE_REF(san_found);
+
+    if (current_name->type == GEN_DNS || current_name->type == GEN_URI) {
+        *san_found = true;
+
+        const char *name = (const char *) ASN1_STRING_data(current_name->d.ia5);
+        size_t name_len = (size_t) ASN1_STRING_length(current_name->d.ia5);
+
+        RESULT_ENSURE(conn->verify_host_fn(name, name_len, conn->data_for_verify_host), S2N_ERR_CERT_UNTRUSTED);
+
+        return S2N_RESULT_OK;
+    }
+
+    if (current_name->type == GEN_IPADD) {
+        *san_found = true;
+
+        /* try to validate an IP address if it's in the subject alt name. */
+        const unsigned char *ip_addr = current_name->d.iPAddress->data;
+        size_t ip_addr_len = (size_t) current_name->d.iPAddress->length;
+
+        RESULT_STACK_BLOB(address, INET6_ADDRSTRLEN + 1, INET6_ADDRSTRLEN + 1);
+
+        if (ip_addr_len == 4) {
+            RESULT_GUARD(s2n_inet_ntop(AF_INET, ip_addr, &address));
+        } else if (ip_addr_len == 16) {
+            RESULT_GUARD(s2n_inet_ntop(AF_INET6, ip_addr, &address));
+        } else {
+            /* we aren't able to parse this value so skip it */
+            return S2N_RESULT_ERROR;
+        }
+
+        /* strlen should be safe here since we made sure we were null terminated AND that inet_ntop succeeded */
+        const char *name = (const char *) address.data;
+        size_t name_len = strlen(name);
+
+        RESULT_ENSURE(conn->verify_host_fn(name, name_len, conn->data_for_verify_host), S2N_ERR_CERT_UNTRUSTED);
+
+        return S2N_RESULT_OK;
+    }
+
+    /* we don't understand this entry type so skip it */
+    RESULT_BAIL(S2N_ERR_CERT_UNTRUSTED);
+}
+
+static S2N_RESULT s2n_verify_host_information_san(struct s2n_connection *conn, X509 *public_cert, bool *san_found)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(public_cert);
+    RESULT_ENSURE_REF(san_found);
+
+    *san_found = false;
+
+    DEFER_CLEANUP(STACK_OF(GENERAL_NAME) *names_list = NULL, GENERAL_NAMES_free_pointer);
+    names_list = X509_get_ext_d2i(public_cert, NID_subject_alt_name, NULL, NULL);
+    RESULT_ENSURE(names_list, S2N_ERR_CERT_UNTRUSTED);
+
+    int n = sk_GENERAL_NAME_num(names_list);
+    RESULT_ENSURE(n > 0, S2N_ERR_CERT_UNTRUSTED);
+
+    for (int i = 0; i < n; i++) {
+        GENERAL_NAME *current_name = sk_GENERAL_NAME_value(names_list, i);
+
+        /* return success on the first entry that passes verification */
+        s2n_result result = s2n_verify_host_information_san_entry(conn, current_name, san_found);
+        if (s2n_result_is_ok(result)) {
+            return S2N_RESULT_OK;
+        }
+    }
+
+    /* if an error was set by one of the entries, then just propagate the error from the last SAN entry call */
+    RESULT_DEBUG_ENSURE(s2n_errno, S2N_ERR_SAFETY); /* make sure we actually set an error on the last call */
```

Comment:
I'm not a fan of making any decisions on s2n_errno not immediately preceded by verifying that an error happened.

At this point, are we sure that the last call to s2n_verify_host_information_san_entry failed?
* If it could have passed (but I don't think this is possible?) then we could be propagating an s2n_errno set elsewhere and unrelated to our code.
* If it must have failed, then can we increase the scope of result and do RESULT_GUARD(result)? That way it's very clear that we're only taking this action for a known failure.